README.md
# emdis-status

REST API web service providing EMDIS hub status information from Ester.

This is a sample project without authentication and therefore for private network use only.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

### Prerequisites

Node.js v8+
Ester v3.15+
Firebird v2.5+

### Installing

Clone (or fork) the repository to your local machine, e.g.

```
$ git clone https://github.com/Jammox/emdis-status
```

CD to the project folder and install the project modules.

```
$ npm install
```

You may need to edit *conf/env/development.json* to reflect the location and port of the Firebird server.
It is recommended to add an alias for the EmdisMail database in your Firebird aliases.conf file.

Run the project.

```
$ npm run monitor
```

Return all hub statuses via:

```
http://localhost:3000/status
```

Return an individual hub status (hub will be a 2 character country identifier e.g. WA, DE, FR):

```
http://localhost:3000/status/hub
```

## Running the tests

Once the service is running tests may run via:

```
$ npm test
```

## Deployment

This service does not provide or require authentication and is intended for internal private network use only.

## Built With

* [node.js](https://nodejs.org)
* [express](https://expressjs.com) - Web framework.
* [node-firebird](https://www.npmjs.com/package/node-firebird) - JavaScript Firebird client.

## Authors

* **James A.Fitches** - *Initial work* - [Jammox](https://github.com/Jammox)

## License

This project is licensed under the GPL3.0 License - see the [LICENSE](https://github.com/Jammox/emdis-status/blob/master/LICENSE) file for details.


CHANGELOG
 * ¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬ *
 *
 * emdis-status
 *
 * REST web service providing EMDIS hub connection status written
 * in Node.js
 * 
 * Copyright (c) 2018-2019 James A.Fitches (J^F)
 *
 * ¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬¬ *

Change log
==========
2018/05/24 - 1.0.0 - J^F > Initial version
2019/06/14 - 1.0.1 - J^F > Package and license update
2019/06/17 - 1.0.2 - J^F > Create tests
2019/12/16 - 1.0.3 - J^F > Package update
2019/12/22 - 1.0.4 - J^F > Package update


test/.eslintrc.js
module.exports = {
  "env": {
    "mocha": true
  },
  "rules": {
    "no-unused-vars": [
      "error",
      {
        "varsIgnorePattern": "should|expect"
      }
    ],
    "no-unused-expressions": "off"
  }
};


test/test-api-rest.js
/*
 * test-api-rest.js
 *
 * 2019/06/14 - 1.0.0 - J^F > Initial version
 *
 * https://www.chaijs.com/guide/
 * https://www.chaijs.com/plugins/chai-http/
 *
 */

const chai = require('chai');
const chaiHttp = require('chai-http');

const should = chai.should();
const { expect } = chai;

chai.use(chaiHttp);

const server = 'http://localhost:3000';
const testhub = 'WA';

function validateRestResponse(res, status) {
  res.should.have.status(status);
  res.should.be.json;

  expect(res.body)
    .to.be.an.instanceof(Array)
    .and.to.have.property(0)
    .that.includes.all.keys(['HUB', 'OUTSEQNUM', 'INSEQNUM', 'OUTDATE', 'INDATE', 'LASTRECVDATE', 'LASTSENDDATE', 'ACTIVEHUB']);
}

describe('emdis-status REST api', () => {
  it('should list ALL hub statuses on /status GET', (done) => {
    // make request
    chai.request(server)
      .get('/status')
      .end((err, res) => {
        expect(err).to.be.null;
        validateRestResponse(res, 200);
        done();
      });
  });

  it('should list a SINGLE hub status on /status/<hub id> GET', (done) => {
    // make request
    chai.request(server)
      .get(`/status/${testhub}`)
      .end((err, res) => {
        expect(err).to.be.null;
        validateRestResponse(res, 200);
        done();
      });
  });
});


api/models/emdisStatusModel.js
/*
 * emdisStatusModel.js
 *
 * 2018/05/24 - 1.0.0 - J^F > Initial version
 *
 */

const debug = require('debug')('emdis-status:emdisStatusModel');
const chalk = require('chalk');
const Firebird = require('node-firebird');

// load config
const config = require('../../conf/config');

// get hub status data from Ester
exports.loadData = (req, res) => {
  let query = 'SELECT hub, outseqnum, inseqnum, outdate, indate, lastrecvdate, lastsenddate, activehub FROM emdiscount';
  const hub = (req.params.Hub || '').toUpperCase();

  if (hub) {
    query += ' WHERE UPPER(hub) = ?';
  }

  Firebird.attach(config.dboptions, (err, db) => {
    if (err) {
      debug(`${chalk.redBright('DbException')}: ${err.stack}`);
      return res.status(500).send({ message: 'Internal Error!' });
    }

    return db.query(query, [hub], (err, result) => {
      db.detach();

      if (err) {
        debug(`${chalk.redBright('QueryException')}: ${err.stack}`);
        return res.status(500).send({ message: 'Internal Error!' });
      }

      if (!result.length) {
        return res.status(404).send({ message: 'Not Found' });
      }

      return res.status(200).send(result);
    });
  });
};


emdis-status.js
/*
 * emdis-status
 *
 * Provides EMDIS hub status information from Ester via a web service
 *
 * Copyright (c) 2018 James Fitches (J^F)
 *
 * 2018/05/24 - 1.0.0 - J^F > Initial version
 *
 * URL: http://localhost:port/status
 *
 */

const debug = require('debug')('emdis-status');
const chalk = require('chalk');
const express = require('express');
const bodyParser = require('body-parser');

const routes = require('./api/routes/emdisStatusRoutes');

// load config
const config = require('./conf/config');

// fire up express
const app = express();

// check port setting
if (typeof config.service.port === 'undefined' || !config.service.port) {
  debug(chalk.redBright('Valid port number required.'));
  process.exit(0);
}

// Trust proxy if defined - allows correct X-Forwarded-For header
if (typeof config.service.proxy !== 'undefined' && config.service.proxy) {
  app.set('trust proxy', config.service.proxy);
  debug(`links is being proxied via ${chalk.green(config.service.proxy)}`);
}

// init json body parser
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

// register routes
routes(app);

// ignition on
app.listen(config.service.port, () => debug(
  `emdis-status is listening on port ${chalk.green(config.service.port)}`
));
